lib/Segments/WooCommerce.php

```diff
@@ -0,0 +1,279 @@
+<?php
+namespace MailPoet\Segments;
+
+use MailPoet\Models\ModelValidator;
+use MailPoet\Models\Subscriber;
+use MailPoet\Models\Segment;
+use MailPoet\Models\SubscriberSegment;
+use MailPoet\Newsletter\Scheduler\Scheduler;
+use MailPoet\Settings\SettingsController;
+use MailPoet\Subscribers\Source;
+
+if(!defined('ABSPATH')) exit;
+
+class WooCommerce {
+  /** @var SettingsController */
+  private $settings;
+
+  function __construct(SettingsController $settings) {
+    $this->settings = $settings;
+  }
+
+  function synchronizeRegisteredCustomer($wp_user_id, $current_filter = null) {
+    if(!$current_filter && !$this->settings->get('enable_wc_hooks_testing')) {
+      return false; // temporarily disable hooks (except for testing)
+    }
+
+    $wc_segment = Segment::getWooCommerceSegment();
+
+    if($wc_segment === false) return;
+
+    $current_filter = $current_filter ?: current_filter();
+    switch($current_filter) {
+      case 'woocommerce_delete_customer':
+        // subscriber should be already deleted in WP users sync
+        $this->unsubscribeUsersFromSegment(); // remove leftover association
+        break;
+      case 'woocommerce_new_customer':
+        $new_customer = true;
+      case 'woocommerce_update_customer':
+      default:
+        $wp_user = \get_userdata($wp_user_id);
+        $subscriber = Subscriber::where('wp_user_id', $wp_user_id)
+          ->findOne();
+
+        if($wp_user === false || $subscriber === false) {
+          // registered customers should exist as WP users and WP segment subscribers
+          return false;
+        }
+
+        $data = array(
+          'is_woocommerce_user' => 1,
+        );
+        if(!empty($new_customer)) {
+          $data['status'] = Subscriber::STATUS_SUBSCRIBED;
+          $data['source'] = Source::WOOCOMMERCE_USER;
+        }
+        $data['id'] = $subscriber->id();
+        $data['deleted_at'] = null; // remove the user from the trash
+
+        $subscriber = Subscriber::createOrUpdate($data);
+        if($subscriber->getErrors() === false && $subscriber->id > 0) {
+          // add subscriber to the WooCommerce Customers segment
+          SubscriberSegment::subscribeToSegments(
+            $subscriber,
+            array($wc_segment->id)
+          );
+        }
+        break;
+    }
+
+    return true;
+  }
+
+  function synchronizeGuestCustomer($order_id, $current_filter = null) {
+    if(!$current_filter && !$this->settings->get('enable_wc_hooks_testing')) {
+      return false; // temporarily disable hooks (except for testing)
+    }
+
+    $wc_order = \get_post($order_id);
+    $wc_segment = Segment::getWooCommerceSegment();
+
+    if($wc_order === false or $wc_segment === false) return;
+
+    $current_filter = $current_filter ?: current_filter();
+    switch($current_filter) {
```

Comment:
Why do we need this `switch` here? We do the step always.

Reply:
I wasn't sure I've used all possible hooks for syncing guests, so I've left this for convenience. I can remove it, but if we discover more cases we'll have to bring it back.

Reply:
I think we should remove it. It is kind of pointless right now and if we discover more cases it wouldn't be very difficult to bring it back.

Reply:
Done. I've left the `current_filter` argument to make the function work in tests.